File: android/measure/src/main/java/sh/measure/android/attributes/AppAttributeProcessor.kt
package sh.measure.android.attributes

import android.content.Context
import sh.measure.android.BuildConfig
import sh.measure.android.utils.PackageInfoProvider

/**
 * Generates the attributes for the app. The attributes include the app version, build version, and
 * the unique ID of the app.
 */
internal class AppAttributeProcessor(
    private val packageInfoProvider: PackageInfoProvider,
    private val context: Context,
) : ComputeOnceAttributeProcessor() {
    private val appVersionKey = "app_version"
    private val appBuildKey = "app_build"
    private val appUniqueIdKey = "app_unique_id"
    private val measureSdkVersion = "measure_sdk_version"

    override fun computeAttributes(): Map<String, Any?> {
        return mapOf(
            appVersionKey to packageInfoProvider.appVersion,
            appBuildKey to packageInfoProvider.getVersionCode(),
            appUniqueIdKey to context.packageName,
            measureSdkVersion to BuildConfig.MEASURE_SDK_VERSION,
        )
    }
}

File: ios/Sources/MeasureSDK/Swift/Attribute/AppAttributeProcessor.swift
//
//  AppAttributeProcessor.swift
//  MeasureSDK
//
//  Created by Adwin Ross on 03/09/24.
//

import Foundation

/// Generates the attributes for the app. The attributes include the app version, build version, and the unique ID of the app.
final class AppAttributeProcessor: BaseComputeOnceAttributeProcessor {
    private lazy var appVersion: String = AttributeConstants.unknown
    private lazy var appBuild: String = AttributeConstants.unknown
    private lazy var appUniqueId: String = AttributeConstants.unknown
    private lazy var measureSdkVersion: String = AttributeConstants.unknown

    override func updateAttribute(_ attribute: Attributes) {
        attribute.appVersion = appVersion
        attribute.appBuild = appBuild
        attribute.appUniqueId = appUniqueId
        attribute.measureSdkVersion = measureSdkVersion
    }

    override func computeAttributes() {
        self.appVersion = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? AttributeConstants.unknown
        self.appBuild = Bundle.main.infoDictionary?["CFBundleVersion"] as? String ?? AttributeConstants.unknown
        self.appUniqueId = Bundle.main.bundleIdentifier ?? AttributeConstants.unknown
        self.measureSdkVersion = FrameworkInfo.version
    }
}
